### ISSUE TITLE
Reasonable limit of new player registrations per IP

### ISSUE THREAD

**Issue by kay27** [minetest/minetest#12072]
## Problem

I received a script to crash any MineClones. The crash will be a separate issue. But what this script does first is creating random user name and registering it on the server.

Player name is always random, 6 characters, and it happens so fast, so I start to worry that it can be a real problem.

What if I just start iterating all possible player names from, say '0', '1', '2', up to 'zzzzzzzzz' and register them all?

I know, IP address will logged, so we can somehow remove these registrations later. But it can be rather uncomfortable for new regular players (not scripts) still who will try to register their accounts at the same time.

Or maybe I miss a something?

## Solutions

It would be nice to have a setup at least of maximum players per single IP.

## Alternatives

Let's discuss?

**Comment by sonolosoft** [comment]
Or max per IP per hour?
- Legit reasons (new name, forgotten, stuck)
- Shared IPs (schools)

**Comment by kay27** [comment]
Well, 10000 should be fine for almost any school, doesn't it?

**Comment by rollerozxa** [comment]
Even if you would have a school server, if it's inside of the school network each student would have their own internal IP address. I'd rather it be set to something really low (5 maybe?) and make it a setting that can be raised if necessary.

**Comment by LoneWolfHT** [comment]
This can be done with a mod, and has already been AFAIK

**Comment by kay27** [comment]
> This can be done with a mod, and has already been AFAIK

True, but it couldn't be a good resolution for large lists, right?

**Comment by appgurueu** [comment]
> > This can be done with a mod, and has already been AFAIK
> 
> True, but it couldn't be a good solution for large lists, right?

Why not? I don't see why a mod would have to be noticeably slower here. You shouldn't be using lists but rather tables - that is, the hash part - to store the IP adresses from which you have already had account registrations. In fact, I believe a hash table plus a heap is pretty much the optimal data structure for this. It can very well be implemented [in Lua](https://the-algorithms.com/de/algorithm/table-heap).

**Comment by sfan5** [comment]
I don't think the engine should do this, partly because I don't think it belongs here but also because there's no single obviously correct way to do this.
Consider the following possible implementations:
* Limit the amount of existing accounts per IP
* Limit the amount of newly registered accounts per IP for a server session
* The above but per some defined time interval
* Limit the amount of existing accounts per IP that are less e.g. 2 weeks old

Not to mention that the suggestion to put the limit at 10000 does not solve the initial problem: If you have 10000 bulk registered accounts you still have a problem on your hands, even if the limit prevented it from being 10001.

**Comment by Festus1965** [comment]
> * Limit the amount of newly registered accounts per IP for a server session

That would be important, as since about 5 days a bad guy/girl is crashing several servers,
sure was my one also (Asia Thailand) with massive attack of same time login attempts.

server can't handle that ...
options yet:
- mod border: close new register
- maybe server password : not tested
- register limit per ip - via mod ipnames, not helping at this stage
